thewalrus/_hafnian.py

```diff
@@ -251,3 +280,48 @@ def hafnian_repeated(A, rpt, mu=None, loop=False, rtol=1e-05, atol=1e-08):
         return haf_rpt_complex(A, nud, mu=mu, loop=loop)
 
     return haf_rpt_real(A, nud, mu=mu, loop=loop)
+
+
+def banded_loophaf(A, loop=False, rtol=1e-05, atol=1e-08):
+    """Returns the loop hafnian of a banded matrix.
+
+    For the derivation see Section V of `'Efficient sampling from shallow Gaussian quantum-optical circuits with local interactions',
+    Qi et al. <https://arxiv.org/abs/2009.11824>`_.
+
+    Args:
+        A (array): a square, symmetric array of even dimensions.
+
+    Returns:
+        np.int64 or np.float64 or np.complex128: the loop hafnian of matrix A.
+    """
+    input_validation(A, atol=atol, rtol=rtol)
+    (n, _) = A.shape
+    w = bandwidth(A)
+    if loop is False:
+        A = A - np.diag(np.diag(A))
+    loop_haf = {(): 1, (1,): A[0, 0]}
+    for t in range(1, n + 1):
+        if t - 2 * w - 1 > 0:
+            lower_end = set(range(1, t - 2 * w))
+        else:
+            lower_end = set()
+        upper_end = set(range(1, t + 1))
+        diff = [item for item in upper_end if item not in lower_end]
+        # Makes sure set ordering is preserved when the difference of two sets is taken
+        # This is also used in the if statement below
+        ps = powerset(diff)
+        lower_end = tuple(lower_end)
+        for D in ps:
+            if lower_end + D not in loop_haf:
+                # pylint: disable=consider-using-generator
+                loop_haf[lower_end + D] = sum(
+                    [
+                        A[i - 1, t - 1]
+                        * loop_haf[
+                            tuple([item for item in lower_end + D if item not in set((i, t))])
+                        ]
+                        for i in D
```

Comment:
It might be more readable to have this with a for loop and a list to which the intermediate results are appended. That might also mean that the `pylint` check can be re-enabled.

Reply:
if you still want to go this way, you could use a `defaultdict` to avoid the `if` statement

Reply:
This format is more similar to how the algorithm looks in the paper where it originates. Might be better to keep as is for people looking at that reference.